File: monitor/dreamhost.py
"""DreamHost (api.dreamhost.com). Key-based cmd API. Docs: api.dreamhost.com."""
from __future__ import annotations

import urllib.parse
import urllib.request
import json

API = "https://api.dreamhost.com/"


def _call(key, cmd, args=None):
    params = {"key": key, "cmd": cmd, "format": "json"}
    params.update(args or {})
    url = API + "?" + urllib.parse.urlencode(params)
    with urllib.request.urlopen(url, timeout=25) as r:
        return json.loads(r.read().decode())


def collect(account, token):
    rows = []
    # Each cmd is best-effort: DreamHost returns {result: ok/error}.
    for cmd, svc, idkey, namekey in (
        ("domain-list_domains", "domain", "domain", "domain"),
        ("dns-list_records", "dns_record", "record", "record"),
    ):
        try:
            data = _call(token, cmd)
        except Exception as e:
            rows.append({
                "provider": "dreamhost", "account": account, "service": "api_error",
                "resource_id": cmd, "name": cmd, "region": "", "status": "error",
                "cost_mtd": 0.0, "tags": {"error": str(e)[:200]}, "raw": {},
            })
            continue
        items = data.get("data", [])
        if isinstance(items, dict):
            items = [items]
        for o in items:
            if not isinstance(o, dict):
                continue
            rid = str(o.get(idkey, o.get("domain", o.get("record", "?"))))
            rows.append({
                "provider": "dreamhost", "account": account, "service": svc,
                "resource_id": f"{cmd}:{rid}", "name": str(o.get(namekey, rid)),
                "region": "", "status": "", "cost_mtd": 0.0,
                "tags": {}, "raw": {k: o.get(k) for k in list(o)[:8]},
            })
    return rows

File: dashboard/lib/monitor/providers/dreamhost.ts
import type { Row } from "../types";

const API = "https://api.dreamhost.com/";

async function call(key: string, cmd: string): Promise<Record<string, never>> {
  const res = await fetch(`${API}?key=${encodeURIComponent(key)}&cmd=${cmd}&format=json`, { cache: "no-store" });
  return res.json();
}

export async function collect(account: string, token: string): Promise<Row[]> {
  const rows: Row[] = [];
  const cmds: [string, string][] = [
    ["domain-list_domains", "domain"],
    ["dns-list_records", "dns_record"],
  ];
  for (const [cmd, service] of cmds) {
    let data: { data?: unknown };
    try {
      data = await call(token, cmd);
    } catch (e) {
      rows.push({
        provider: "dreamhost", account, service: "api_error", resource_id: cmd,
        name: cmd, region: "", status: "error", cost_mtd: 0,
        tags: { error: String(e).slice(0, 200) }, raw: {},
      });
      continue;
    }
    let items = (data.data || []) as unknown;
    if (!Array.isArray(items)) items = [items];
    for (const o of items as Record<string, unknown>[]) {
      if (typeof o !== "object" || !o) continue;
      const rid = String(o.domain || o.record || "?");
      rows.push({
        provider: "dreamhost", account, service, resource_id: `${cmd}:${rid}`,
        name: String(o.domain || o.record || rid), region: "", status: "", cost_mtd: 0,
        tags: {}, raw: Object.fromEntries(Object.entries(o).slice(0, 8)) as Record<string, unknown>,
      });
    }
  }
  return rows;
}
